File: Analyzer/cpp/lib/Manager/DBMgr/DBMgr.cpp
#include "Manager/DBMgr/DBMgr.h"

#include <string>
#include <fstream>
#include <iostream>
#include <spdlog/spdlog.h>

using namespace std;

/*
 * Implementation of DBMgr
 */

DBMgr::DBMgr(string db_path, size_t cache_size) {
	db = NULL;
	if (sqlite3_open(db_path.c_str(), &db) != SQLITE_OK) {
		spdlog::error("DBMgr::DBMgr: Fail to open database: {}", db_path);
	}
	execSimple("PRAGMA journal_mode=WAL;");
	execSimple("PRAGMA foreign_keys=ON;");
	row_cache.clear();
	this->cache_size = cache_size;
}

DBMgr::~DBMgr() {
	sqlite3_close(db);
}

string DBMgr::loadSql(string sql_path) {
	ifstream ifs(sql_path);
	stringstream buffer;
	buffer << ifs.rdbuf();
	return buffer.str();
}

bool DBMgr::execSimple(const string &sql) {
	char *err_msg = nullptr;
	int rc = sqlite3_exec(db, sql.c_str(), nullptr, nullptr, &err_msg);
	if (rc != SQLITE_OK) {
		sqlite3_free(err_msg);
		return false;
	}
	return true;
}

bool DBMgr::bindValue(sqlite3_stmt *stmt, int index, const SqlValue &val) {
	return visit([&](auto &&arg) -> bool {
		using T = decay_t<decltype(arg)>;
		if constexpr (is_same_v<T, nullptr_t>) {
			return sqlite3_bind_null(stmt, index) == SQLITE_OK;
		} else if constexpr (is_same_v<T, int>) {
			return sqlite3_bind_int(stmt, index, arg) == SQLITE_OK;
		} else if constexpr (is_same_v<T, long long>) {
			return sqlite3_bind_int64(stmt, index, arg) == SQLITE_OK;
		} else if constexpr (is_same_v<T, string>) {
			return sqlite3_bind_text(stmt, index, arg.c_str(), -1, SQLITE_TRANSIENT) == SQLITE_OK;
		} else {
			return false;
		}
	}, val);
}

std::string DBMgr::params2str(const SqlParams &sql_params) {
	string result;
	for (auto &param : sql_params) {
		string param_str = std::visit([](const auto &x) -> string {
			using T = decay_t<decltype(x)>;
			if constexpr (is_same_v<T, nullptr_t>) {
				return "NULL";
			} else if constexpr (is_same_v<T, string>) {
				return x;
			} else {
				return to_string(x);
			}
		}, param);
		result = result + param_str + " ";
	}
	return result;
}

bool DBMgr::beginTransaction() {
	return execSimple("BEGIN TRANSACTION;");
}

bool DBMgr::commit() {
	return execSimple("COMMIT;");
}

bool DBMgr::rollback() {
	return execSimple("ROLLBACK;");
}

bool DBMgr::execSql(const std::string &sql, const SqlParams &params) {
	if (cache_size) {
		if (row_cache.size() < cache_size) {
			row_cache.push_back(params);
			// FIXME: Return true without other checking may be wrong.
			return true;
		} else {
			bool ret = execSqlBatch(sql, row_cache);
			// FIXME: Clear whatever ret is true.
			row_cache.clear();
			return ret;
		}
	} 
	sqlite3_stmt *stmt = nullptr;
	if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, nullptr) != SQLITE_OK) {
		spdlog::error("DBMgr::execSql: Prepare failed.");
		return false;
	}
	for (size_t param_idx = 0; param_idx < params.size(); param_idx++) {
		if (!bindValue(stmt, param_idx + 1, params[param_idx])) {
			spdlog::error("DBMgr::execSql: Bind failed at param {}", param_idx + 1);
			sqlite3_finalize(stmt);
			return false;
		}
	}
	int rc = sqlite3_step(stmt);
	if (rc != SQLITE_DONE) {
		spdlog::error("DBMgr::execSql: Execution failed.");
		sqlite3_finalize(stmt);
		return false;
	} else {
		int changes = sqlite3_changes(db);
		if (changes == 0) {
			//spdlog::warn("DBMgr::execSql: Repeated insert:\n {}", params2str(params));
		}
	}
	sqlite3_finalize(stmt);
	return true;
}

bool DBMgr::execSqlBatch(const std::string &sql, const std::vector<SqlParams> &rows) {
	if (rows.empty()) {
		return true;
	}

	sqlite3_stmt *stmt = nullptr;

	if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, nullptr) !=SQLITE_OK) {
		spdlog::error("DBMgr::execSqlBatch: Prepare failed.");
		return false;
	}

	if (!beginTransaction()) {
		sqlite3_finalize(stmt);
		return false;
	}

	bool success = true;
	for (const auto &row : rows) {
		sqlite3_reset(stmt);
		sqlite3_clear_bindings(stmt);

		for (size_t param_idx = 0; param_idx < row.size(); param_idx++) {
			if (!bindValue(stmt, param_idx + 1, row[param_idx])) {
				spdlog::error("DBMgr::execSqlBatch: Bind failed at param {}", param_idx + 1);
				success = false;
				break;
			}
		}
		if (!success) {
			break;
		}

		int rc = sqlite3_step(stmt);
		if (rc != SQLITE_DONE) {
			spdlog::error("DBMgr::execSqlBatch: Execution failed.");
			success = false;
			break;
		}
	}
	if (success) {
		if (!commit()) {
			success = false;
		}
	} else {
		rollback();
	}
	sqlite3_finalize(stmt);
	return success;
}

sqlite3_stmt *DBMgr::prepareQuery(const string &sql, const SqlParams &params) {
	sqlite3_stmt *stmt = nullptr;
	if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, nullptr) != SQLITE_OK) {
		spdlog::error("DBMgr::prepareQuery: Prepare query failed.");
		return nullptr;
	}
	for (size_t param_idx = 0; param_idx < params.size(); param_idx++) {
		if (!bindValue(stmt, param_idx + 1, params[param_idx])) {
			spdlog::error("DBMgr::prepareQuery: Bind query param failed at {}", param_idx + 1);
			sqlite3_finalize(stmt);
			return nullptr;
		}
	}
	return stmt;
}

void DBMgr::finalizeQuery(sqlite3_stmt *stmt) {
	if (stmt) {
		sqlite3_finalize(stmt);
	}
}

File: Analyzer/python/RaceMinerLib/Manager/DBMgr/DBMgr.py
import sqlite3
import logging
from typing import List, Any, Optional

SqlValue = Any
SqlParams = List[SqlValue]

class DBMgr:
    def __init__(self, db_path: str, cache_size: int = 0):
        self.logger = logging.getLogger("DBMgr")
        if not self.logger.handlers:
            handler = logging.StreamHandler()
            formatter = logging.Formatter(
                "[%(asctime)s][%(levelname)s] %(message)s"
            )
            handler.setFormatter(formatter)
            self.logger.addHandler(handler)
            self.logger.setLevel(logging.INFO)

        try:
            self.conn = sqlite3.connect(db_path)
            self.conn.execute("PRAGMA journal_mode=WAL;")
            self.conn.execute("PRAGMA foreign_keys=ON;")
            self.logger.info(f"Open database: {db_path}")
        except Exception as e:
            self.logger.error(f"Failed to open database: {e}")
            raise

        self.cache_size = cache_size
        self.row_cache: List[SqlParams] = []

    def close(self):
        try:
            self.conn.close()
            self.logger.info("Database closed")
        except Exception as e:
            self.logger.error(f"Close failed: {e}")

    def exec_simple(self, sql: str) -> bool:
        try:
            self.conn.execute(sql)
            self.conn.commit()
            self.logger.debug(f"Exec simple SQL: {sql}")
            return True
        except Exception as e:
            self.logger.error(f"[exec_simple] {e} | SQL: {sql}")
            return False

    def exec_sql(self, sql: str, params: SqlParams) -> bool:
        if self.cache_size > 0:
            self.row_cache.append(params)

            if len(self.row_cache) < self.cache_size:
                self.logger.debug(f"Cache row: {params}")
                return True
            else:
                self.logger.info("Cache full, flushing batch...")
                ret = self.exec_sql_batch(sql, self.row_cache)
                self.row_cache.clear()
                return ret

        try:
            cur = self.conn.execute(sql, params)
            self.conn.commit()

            if cur.rowcount == 0:
                self.logger.debug(f"No change: {params}")

            return True
        except Exception as e:
            self.logger.error(f"[exec_sql] {e} | SQL: {sql} | Params: {params}")
            return False

    def exec_sql_batch(self, sql: str, rows: List[SqlParams]) -> bool:
        if not rows:
            return True

        try:
            cur = self.conn.cursor()
            self.conn.execute("BEGIN TRANSACTION;")

            for idx, row in enumerate(rows):
                try:
                    cur.execute(sql, row)
                except Exception as e:
                    self.logger.error(f"[batch row {idx}] {e} | Params: {row}")
                    raise

            self.conn.commit()
            self.logger.info(f"Batch executed: {len(rows)} rows")
            return True

        except Exception as e:
            self.logger.error(f"[exec_sql_batch] rollback due to error: {e}")
            self.conn.rollback()
            return False

    def prepare_query(self, sql: str, params: SqlParams):
        try:
            cur = self.conn.cursor()
            cur.execute(sql, params)
            self.logger.debug(f"Query prepared: {sql} | {params}")
            return cur
        except Exception as e:
            self.logger.error(f"[prepare_query] {e} | SQL: {sql} | Params: {params}")
            return None

    def finalize_query(self, cursor):
        if cursor:
            cursor.close()

    def begin_transaction(self):
        try:
            self.conn.execute("BEGIN TRANSACTION;")
            self.logger.debug("Begin transaction")
        except Exception as e:
            self.logger.error(f"Begin transaction failed: {e}")

    def commit(self):
        try:
            self.conn.commit()
            self.logger.debug("Commit")
        except Exception as e:
            self.logger.error(f"Commit failed: {e}")

    def rollback(self):
        try:
            self.conn.rollback()
            self.logger.debug("Rollback")
        except Exception as e:
            self.logger.error(f"Rollback failed: {e}")

    def flush(self, sql: str):
        if self.row_cache:
            self.logger.info(f"Flushing remaining {len(self.row_cache)} rows...")
            self.exec_sql_batch(sql, self.row_cache)
            self.row_cache.clear()
